File: 동적계획법1/지우/BOJ_2839_설탕 배달.cpp
#include <iostream>
#include <vector>

using namespace std;
int INF = 98765432;

int main() {
    int N; cin >> N;
    vector<int> dp(5001, INF);

    for(int i=3; i<=5000; i+=3) {
        dp[i] = i/3;
    }
    for(int i=5; i<=5000; i+=5) {
        dp[i] = min(dp[i], i/5);
    }

    for(int t=3; t<=5000; t+=3) {
        for(int f=5; f<=5000; f+=5) {
            if(t+f > 5000) break;
            dp[t+f] = min(dp[t+f], dp[t]+dp[f]);
        }
    }

    if(dp[N] == INF) cout << -1;
    else cout << dp[N];
    
    return 0;
}

File: 동적계획법1/나영/BOJ_2839_설탕 배달.py
import sys
input = sys.stdin.readline

n = int(input())
inf = float('inf')
dp = [inf] * (n+1)

if n >= 3 : dp[3] = 1
if n >= 5 : dp[5] = 1

for i in range(6, n+1) : 
    dp[i] = min(dp[i-3], dp[i-5]) + 1

print(dp[n] if dp[n] != inf else -1)